framework/stats/vulkan_stats_provider.cpp

```diff
@@ -0,0 +1,460 @@
+		uint32_t    ctr_idx, div_idx;
+
+		std::regex name_regex(init.name);
+		std::regex div_regex(init.divisor_name);
+
+		for (uint32_t i = 0; !(found_ctr && found_div) && i < descs.size(); i++)
+		{
+			if (!found_ctr && std::regex_match(descs[i].name, name_regex))
+			{
+				ctr_idx   = i;
+				found_ctr = true;
+			}
+			if (!found_div && std::regex_match(descs[i].name, div_regex))
+			{
+				div_idx   = i;
+				found_div = true;
+			}
+		}
+
+		if (found_ctr && found_div)
+		{
+			if ((descs[ctr_idx].flags & VK_PERFORMANCE_COUNTER_DESCRIPTION_PERFORMANCE_IMPACTING_KHR) ||
+			    (init.divisor_name != "" && descs[div_idx].flags != VK_PERFORMANCE_COUNTER_DESCRIPTION_PERFORMANCE_IMPACTING_KHR))
+			{
+				performance_impact = true;
+			}
+
+			// Record the counter data
+			counter_indices.emplace_back(ctr_idx);
+			if (init.divisor_name == "")
+			{
+				stat_data[index] = StatData(ctr_idx, counters[ctr_idx].storage);
+			}
+			else
+			{
+				counter_indices.emplace_back(div_idx);
+				stat_data[index] = StatData(ctr_idx, counters[ctr_idx].storage, init.scaling,
+				                            div_idx, counters[div_idx].storage);
+			}
+		}
+	}
+
+	if (performance_impact)
+		LOGW("The collection of performance counters may impact performance");
+
+	if (counter_indices.size() == 0)
+		return;        // No stats available
+
+	// Acquire the profiling lock, without which we can't collect stats
+	VkAcquireProfilingLockInfoKHR info{};
+	info.sType   = VK_STRUCTURE_TYPE_ACQUIRE_PROFILING_LOCK_INFO_KHR;
+	info.timeout = 2000000000;        // 2 seconds (in ns)
+
+	if (vkAcquireProfilingLockKHR(device.get_handle(), &info) != VK_SUCCESS)
+	{
+		stat_data.clear();
+		counter_indices.clear();
+		LOGW("Profiling lock acquisition timed-out");
+		return;
+	}
+
+	// Now we know the counters and that we can collect them, make a query pool for the results.
+	if (!create_query_pools(uint32_t(num_framebuffers), queue_family_index))
+	{
+		stat_data.clear();
+		counter_indices.clear();
+		return;
+	}
+
+	// These stats are fully supported by this provider and in a single pass, so remove
+	// from the requested set.
+	// Subsequent providers will then only look for things that aren't already supported.
+	for (const auto &s : stat_data)
+		requested_stats.erase(s.first);
+}
+
+VulkanStatsProvider::~VulkanStatsProvider()
+{
+	if (stat_data.size() > 0)
+	{
+		// Release profiling lock
+		vkReleaseProfilingLockKHR(device.get_handle());
+	}
+}
+
+bool VulkanStatsProvider::fill_vendor_data()
+{
+	const auto &pd_props = device.get_gpu().get_properties();
+	if (pd_props.vendorID == 0x14E4)        // Broadcom devices
+	{
+		LOGI("Using Vulkan performance counters from Broadcom device");
+
+		// NOTE: The names here are actually regular-expressions.
+		// Counter names can change between hardware variants for the same vendor,
+		// so regular expression names mean that multiple h/w variants can be easily supported.
+		// clang-format off
```

Comment:
why do we turn clang-format off in a few places? how does the output look like, and can we fix it by adjusting the .clang-format file?

Reply:
I've turned off clang-formatting for a few places where tables are being defined in the code.
Without doing so the table is just tightly aligned which makes it much harder to read in my opinion:
```
		vendor_data = {
		    {StatIndex::gpu_cycles, {"cycle_count"}},
		    {StatIndex::gpu_vertex_cycles, {"qpu_active_cycles_vertex_coord_user"}},
		    {StatIndex::gpu_fragment_cycles, {"qpu_active_cycles_fragment"}},
		    {StatIndex::gpu_fragment_jobs, {"render_jobs_completed"}},
		    {StatIndex::gpu_ext_reads, {"gpu_mem_reads"}},
		    {StatIndex::gpu_ext_writes, {"gpu_mem_writes"}},
		    {StatIndex::gpu_ext_read_bytes, {"gpu_bytes_read"}},
		    {StatIndex::gpu_ext_write_bytes, {"gpu_bytes_written"}},
		};
```
compared with:
```
		vendor_data = {
		    {StatIndex::gpu_cycles,          {"cycle_count"}},
		    {StatIndex::gpu_vertex_cycles,   {"qpu_active_cycles_vertex_coord_user"}},
		    {StatIndex::gpu_fragment_cycles, {"qpu_active_cycles_fragment"}},
		    {StatIndex::gpu_fragment_jobs,   {"render_jobs_completed"}},
		    {StatIndex::gpu_ext_reads,       {"gpu_mem_reads"}},
		    {StatIndex::gpu_ext_writes,      {"gpu_mem_writes"}},
		    {StatIndex::gpu_ext_read_bytes,  {"gpu_bytes_read"}},
		    {StatIndex::gpu_ext_write_bytes, {"gpu_bytes_written"}},
		};
```
There is no way to do this in the clang-format file as far as I'm aware.

Reply:
Yes I agree, thanks!